services/appflowy-collaborate/src/collab/cache/collab_cache.rs

```diff
@@ -196,48 +210,74 @@ impl CollabCache {
       Err(AppError::RecordNotFound(_)) => (Rid::default(), None),
       Err(err) => return Err(err),
     };
+
     let updates = self
       .get_workspace_updates(workspace_id, Some(&object_id), Some(rid), None)
       .await?;
-    tracing::trace!("replaying {} updates for {}", updates.len(), object_id);
+
+    // Determine the state vector to use for generating the diff:
+    // 1. If provided in the query, use it directly
+    // 2. Otherwise, try to decode from the existing encoded collab
+    // 3. Fall back to empty state vector if decoding fails or no collab exists
+    let from = from.unwrap_or_else(|| {
+      encoded_collab
+        .as_ref()
+        .and_then(|ec| match ec.version {
+          EncoderVersion::V1 => StateVector::decode_v1(&ec.state_vector).ok(),
+          EncoderVersion::V2 => StateVector::decode_v2(&ec.state_vector).ok(),
+        })
+        .unwrap_or_default()
+    });
+
     if !updates.is_empty() {
-      let mut collab = match encoded_collab {
-        Some(encoded_collab) => {
-          let options = CollabOptions::new(object_id.to_string(), default_client_id())
-            .with_data_source(match encoded_collab.version {
-              EncoderVersion::V1 => DataSource::DocStateV1(encoded_collab.doc_state.into()),
-              EncoderVersion::V2 => DataSource::DocStateV2(encoded_collab.doc_state.into()),
-            });
-          Collab::new_with_options(CollabOrigin::Server, options)
-            .map_err(|err| AppError::Internal(err.into()))?
-        },
-        None => {
-          let options = CollabOptions::new(object_id.to_string(), default_client_id());
-          Collab::new_with_options(CollabOrigin::Server, options)
-            .map_err(|err| AppError::Internal(err.into()))?
-        },
-      };
-      {
-        let mut tx = collab.transact_mut();
-        for msg in updates {
-          if msg.object_id == object_id {
-            let update = match msg.update_flags {
-              UpdateFlags::Lib0v1 => Update::decode_v1(&msg.update),
-              UpdateFlags::Lib0v2 => Update::decode_v2(&msg.update),
+      encoded_collab = self
+        .thread_pool
+        .install(|| {
```

Comment:
What I'm worried about is that for small collabs (ie. DatabaseRows), process of calling install may take longer than the decoding+encoding. Maybe we could add some metrics to thing call?

Reply:
Yes. I will fallback to previous approach.